src/db/model/sync.rs

```diff
@@ -17,37 +19,89 @@ use crate::db::{model::tangle::MilestoneIndex, MongoDb};
+}
+
 impl MongoDb {
+    /// If available, returns the [`SyncRecord`] associated with the provided milestone `index`.
+    pub async fn get_sync_record_by_index(&self, index: MilestoneIndex) -> Result<Option<SyncRecord>, Error> {
+        self.0
+            .collection::<SyncRecord>(SyncRecord::COLLECTION)
+            .find_one(doc! {"milestone_index": index}, None)
+            .await
+    }
+
     /// Upserts a [`SyncRecord`] to the database.
-    pub async fn upsert_sync_record(&self, record: &SyncRecord) -> Result<UpdateResult, Error> {
+    pub async fn upsert_sync_record(&self, index: MilestoneIndex) -> Result<UpdateResult, Error> {
         self.0
             .collection::<SyncRecord>(SyncRecord::COLLECTION)
             .update_one(
-                doc! {"milestone_index": record.milestone_index},
-                doc! {"$set": bson::to_document(record)?},
+                doc! {"_id": index},
+                doc! {"$set": bson::to_document(&SyncRecord{milestone_index: index})?},
                 UpdateOptions::builder().upsert(true).build(),
             )
             .await
     }
     /// Retrieves the sync records sorted by [`milestone_index`](SyncRecord::milestone_index).
-    pub async fn sync_records_sorted(&self) -> Result<impl Stream<Item = Result<SyncRecord, Error>>, Error> {
+    pub async fn sync_records_sorted(
+        &self,
+        range: RangeInclusive<MilestoneIndex>,
+    ) -> Result<impl Stream<Item = Result<SyncRecord, Error>>, Error> {
         self.0
             .collection::<SyncRecord>(SyncRecord::COLLECTION)
             .find(
-                doc! { "synced": true },
-                FindOptions::builder().sort(doc! {"milestone_index": 1u32}).build(),
+                doc! { "milestone_index": { "$gte": range.start(), "$lte": range.end() } },
+                FindOptions::builder().sort(doc! {"milestone_index": 1}).build(),
             )
             .await
     }
+
+    /// Retrieves a [`SyncData`] structure that contains the completed and gaps ranges.
+    pub async fn get_sync_data(&self, range: RangeInclusive<MilestoneIndex>) -> Result<SyncData, Error> {
+        let mut res = self.sync_records_sorted(range.clone()).await?;
+        let mut sync_data = SyncData::default();
+        let mut last_record: Option<MilestoneIndex> = None;
+        while let Some(SyncRecord { milestone_index }) = res.try_next().await? {
+            // Missing records go into gaps
+            if let Some(&last) = last_record.as_ref() {
+                if last + 1 != milestone_index {
+                    sync_data.gaps.push(last + 1..=milestone_index - 1);
+                }
+            } else if *range.start() < milestone_index {
+                sync_data.gaps.push(*range.start()..=milestone_index - 1)
+            }
+            match sync_data.completed.last_mut() {
+                Some(last) => {
+                    if last.end + 1 == milestone_index {
+                        last.end += 1;
```

Comment:
Sorry but that's weird to me. Where else is `last.end` modified, and when it's mutated here as well, is that taken into account in the other place?

Reply:
alright, I get it now!